pkg/network/ebpf/c/protocol-classification-helpers.h

```diff
@@ -0,0 +1,214 @@
+        bpf_skb_load_bytes(skb, offset, buf, 15);
+    } else if (offset + 13 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 14);
+    } else if (offset + 12 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 13);
+    } else if (offset + 11 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 12);
+    } else if (offset + 10 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 11);
+    } else if (offset + 9 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 10);
+    } else if (offset + 8 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 9);
+    } else if (offset + 7 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 8);
+    } else if (offset + 6 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 7);
+    } else if (offset + 5 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 6);
+    } else if (offset + 4 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 5);
+    } else if (offset + 3 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 4);
+    } else if (offset + 2 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 3);
+    } else if (offset + 1 < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 2);
+    } else if (offset < len) {
+        bpf_skb_load_bytes(skb, offset, buf, 1);
+    }
+}
+
+// checks if we have seen that tcp packet before. If can happen if a packet travels multiple interfaces or retransmissions.
+static __always_inline bool has_sequence_seen_before(conn_tuple_t *tup, skb_info_t *skb_info) {
+    if (!skb_info || !skb_info->tcp_seq) {
+        return false;
+    }
+
+    u32 *tcp_seq = bpf_map_lookup_elem(&connection_states, tup);
+
+    // check if we've seen this TCP segment before. this can happen in the
+    // context of localhost traffic where the same TCP segment can be seen
+    // multiple times coming in and out from different interfaces
+    if (tcp_seq != NULL && *tcp_seq == skb_info->tcp_seq) {
+        return true;
+    }
+
+    bpf_map_update_elem(&connection_states, tup, &skb_info->tcp_seq, BPF_ANY);
+    return false;
+}
+
+// A shared implementation for the runtime & prebuilt socket filter that classifies the protocols of the connections.
+static __always_inline void protocol_classifier_entrypoint(struct __sk_buff *skb) {
+    skb_info_t skb_info = {0};
+    conn_tuple_t tup = {0};
+
+    if (!read_conn_tuple_skb(skb, &skb_info, &tup)) {
+        return;
+    }
+
+    if (!should_process_packet(skb, &skb_info, &tup)) {
+        return;
+    }
+
+    if (has_sequence_seen_before(&tup, &skb_info)) {
+        return;
+    }
+
+    char request_fragment[CLASSIFICATION_MAX_BUFFER];
+    bpf_memset(request_fragment, 0, sizeof(request_fragment));
+    read_into_buffer_skb_post_4_5_0((char *)request_fragment, skb, &skb_info);
+
+    struct sock **protocol_key_ptr = bpf_map_lookup_elem(&conn_tuple_to_socket_map, &tup);
+    if (protocol_key_ptr == NULL) {
+        log_debug("[protocol_classifier_entrypoint]: protocol_key_ptr is null, aborting\n");
```

Comment:
that's a debug log, and i'm not sure it will give us anything
furthermore, we can pass max 3 parameters to log_debug, and conn_tuple has 9 fields, meaning we will have to perform 3 calls to log_debug which won't make it easier to read